### DISCUSSION TITLE
A Guide for building community prisma generators

### DISCUSSION THREAD

**Discussion by YassinEldeeb** [prisma/prisma#10721] [Q&A] [answered]
Currently the only way for the community to build amazing prisma generators is to look for the already built generators code to see how to use `@prisma/generator-helper` and the `@prisma/sdk` cause there's zero resources about them neither officially nor from the community.

So I would like to write a guide on how to get started with developing a prisma generator so we can eliminate the struggle of developers to start developing their amazing ideas cause I find it extremely demotivating when I started to make [my own generator](https://github.com/YassinEldeeb/prisma-tgql-types-gen) when I couldn't find a single resource on how to get started.

so what I would like to create is:
1. Make a little CLI tool for people to use to setup all of the boilerplate needed to start developing a new prisma generator which will contain:
- Private yarn workspace for testing the generator.
- Two projects in the workspace: 
   - Generator package.
   - Simple Node.js project that has prisma already setup with a sample definations with the generator package symlink-ed in `package.json` that a user can test running `prisma generate` to see the generated output.

2. Illustrate the different concepts they'll come across when developing prisma generators like `dmmf`, `generator configs` ..etc
3. Illustrate the power of using `@prisma/generator-helper` which will help saving a lot of time and effort.
4. Illustrate the basic usage for `@prisma/sdk`.
5. Publishing the package to npm.
6. Explaining how to write tests for the generator.

**I would like to hear your opinion about this idea 🤗**

**The question is:**
where would you like to put this guide? in a `.md` file on this repo just like `CONTRIBUTING.md` or on prisma/docs in an mdx file to be on [prisma.io](prisma.io) or perhaps somewhere else 🤔

and can the CLI tool for setting up the boilerplate be merged with the prisma ecosystem or It must be a community maintained thing and if yes where should it be placed?

**Comment by nikolasburk** [comment, marked as answer]
Hey Yassin 👋 this is a great initiative and I'd love to see such a guide that makes it easier for the community to build their own generators!!

However, there's a major caveat to this. The DMMF is an undocumented API that's _not_ considered part of the official API surface and might break unexpectedly with any release, so it's very intentionally not documented in our own resources (somewhat related context also in this issue: https://github.com/prisma/prisma/issues/10725). 

I would still love to see such a guide, do you maybe have a personal blog where you could publish it? Otherwise, platforms like [Hashnode](https://hashnode.com/) or [dev.to](https://dev.to/) are great to publish blog posts without the need of setting up your own blog. While we are not able to include such a guide in our own resources, I do think we can link to it from the [Community generators](https://www.prisma.io/docs/concepts/components/prisma-schema/generators#community-generators)-section in our docs!

**Reply by YassinEldeeb** [reply]
Thank you for reaching out @nikolasburk you're really helpful here and on Twitter 😊

Yeah I can write a blog on [dev.to/yassineldeeb](https://dev.to/yassineldeeb) explaining all of that + making that `create-prisma-generator` CLI for setting up all of the needed boilerplate.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
